### PR TITLE
#99 Support deep links

### LINKED ISSUE #78: Wrappers over AndroidDriver commands in Selenide-Appium
Hi @asolntsev,
May I ask you please to help me with clarification of the following question:
Does Selenide-Appium have methods/wrappers to work with Android native keyboard? The one that can be accessed through AndroidDriver instance like:
`(WebDriverRunner.getWebDriver() as AndroidDriver).pressKey(KeyEvent(DIGIT_0))`
If not maybe there are some plans in the nearest future to implement such wrappers in Selenide-Appium?
Thanks a lot in advance for the help.

### LINKED ISSUE #99: Deeplinks support
Allow to open deep links on ios\android.

### PR DESCRIPTION
**Pull request by BorisOsipov**
Resolves #99 
Resolves #78
+
Support terminate app
Replace Android calculator test - there is not calculator on some emulator images.

- [ ] Checkstyle and unit tests are passed locally with my changes by running `gradlew check` command
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/appium/SelenideAppium.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+   * @param appId - applicationId for Android or bundleId for iOS
+   */
+  public static void terminateApp(String appId) {
+    appiumNavigator.terminateApp(AppiumDriverRunner.getMobileDriver(), appId);
+  }
+
+  /**
+   * Navigate app back to previous screen
+   */
+  public static void back() {
```
Comment (amuthansakthivel) [inline src/main/java/com/codeborne/selenide/appium/SelenideAppium.java:57]: @BorisOsipov - I think, we have to tweak it a bit.

if( androidDriver()) we can use use back
if(iosdriver() we have to throw exception as back is not supported in ios

File: src/test/java/com/codeborne/selenide/appium/AppiumNavigatorTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.mockito.ArgumentCaptor;
+
+class AppiumNavigatorTest {
+  private final AppiumNavigator appiumNavigator = new AppiumNavigator();
+
+  @BeforeEach
+  @AfterEach
+  void resetSettings() {
+    Configuration.browserSize = "1024x768";
+    Configuration.pageLoadTimeout = 30_000;
```
Comment (amuthansakthivel) [inline src/test/java/com/codeborne/selenide/appium/AppiumNavigatorTest.java:25]: @BorisOsipov - Sorry for being lame, 30_000 (Why there is an underscore)

File: src/main/java/com/codeborne/selenide/appium/AppiumDriverRunner.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  @Nonnull
+  @CheckReturnValue
+  @SuppressWarnings("unchecked")
+  public static <T extends AppiumDriver> T getMobileDriver() {
+    if (isAndroidDriver() || isIosDriver()) {
+      return isAndroidDriver() ? (T) getAndroidDriver() : (T) getIosDriver();
+    }
+    throw new ClassCastException("WebDriver is not instance of AndroidDriver or IOSDriver");
+  }
+
```
Comment (amuthansakthivel) [inline src/main/java/com/codeborne/selenide/appium/AppiumDriverRunner.java:50]: Thanks for adding java docs @BorisOsipov

File: src/main/java/com/codeborne/selenide/appium/SelenideAppium.java (hunk 1 above)
Comment (BorisOsipov) [reply src/main/java/com/codeborne/selenide/appium/SelenideAppium.java:57]: >back is not supported in ios

It is supported but only for web context or screens with native NavBar component.
I think we should not throw error but maybe add info to javadoc.
https://github.com/appium/appium-xcuitest-driver/blob/master/lib/commands/navigation.js#L11-L38

Wdyt?

Comment (amuthansakthivel) [reply src/main/java/com/codeborne/selenide/appium/SelenideAppium.java:57]: yes @BorisOsipov you are correct. It is supported in web view and hence we cannot throw any exception.

File: src/test/java/com/codeborne/selenide/appium/DeepLinkLauncherTest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import io.appium.java_client.android.AndroidDriver;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+
+class DeepLinkLauncherTest {
+  DeepLinkLauncher deepLinkLauncher = new DeepLinkLauncher();
+
+  @Test
+  void shouldSendAppIdForAndroid() {
```
[the remark this answers is not in the document]
Comment (BorisOsipov) [reply src/test/java/com/codeborne/selenide/appium/DeepLinkLauncherTest.java:16]: Why do you think the new name will be better? FMPOV here we test that our implementation call execScript with correct parameters for ios\android cases.

[the remark this answers is not in the document]
Comment (amuthansakthivel) [reply src/test/java/com/codeborne/selenide/appium/DeepLinkLauncherTest.java:16]: @BorisOsipov AppiumNavigatorTest have already a method called shouldSendAppIdForAndroid

[the remark this answers is not in the document]
Comment (BorisOsipov) [reply src/test/java/com/codeborne/selenide/appium/DeepLinkLauncherTest.java:16]: Renamed test names in AppiumNavigatorTest

**Review by amuthansakthivel — APPROVED** [review]
LGTM

File: src/main/java/com/codeborne/selenide/appium/DeepLinkLauncher.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      // then Appium will auto accept the alert that should be shown now.
+      if (!canAutoAcceptAlerts(driver)) {
+        // Wait for the notification and accept it
+        // When using an iOS simulator you will only get the pop-up once, all the other times it won't be shown
+        try {
+          SelenideElement openButton =
+            $(iOSNsPredicateString("type == 'XCUIElementTypeButton' && name CONTAINS 'Open'"));
+          openButton.shouldBe(enabled, SAFARI_ELEMENTS_TIMEOUT).click();
+        } catch (AssertionError e) {
+          // ignore
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/appium/DeepLinkLauncher.java:57]: 1. Maybe we should log this `e`? At lest with `debug` level?
2. Don't we spend 4 seconds and take an unneeded screenshot in this case?
@BorisOsipov

File: src/main/java/com/codeborne/selenide/appium/AppiumNavigator.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  public void launchApp(Supplier<Runnable> driverSupplier) {
+    Configuration.browserSize = null;
+    Configuration.pageLoadTimeout = 0;
+    SelenideLogger.run("launch app", "", driverSupplier.get());
+  }
+
+  public void terminateApp(AppiumDriver driver, String appId) {
+    SelenideLogger.run("terminate app", appId, () -> {
+      HashMap<String, String> params = new HashMap<>();
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/appium/AppiumNavigator.java:25]: I usually declare variable as `Map<String, String> params`, not `HashMap`.

**Review by asolntsev — APPROVED** [review]

**Comment by asolntsev** [comment]
@BorisOsipov Great job, thank you!

File: src/main/java/com/codeborne/selenide/appium/DeepLinkLauncher.java (hunk 5 above)
Comment (BorisOsipov) [reply src/main/java/com/codeborne/selenide/appium/DeepLinkLauncher.java:57]: 1. agree. will fix.
2. we spend 30 secs. That's why I open backdoor [via public timeout](https://github.com/selenide/selenide-appium/blob/7c8ab08ce2820d169978d782fe23826700f22876/src/main/java/com/codeborne/selenide/appium/DeepLinkLauncher.java#L24) to allow user fix it temporary. I have no idea how to disable unneeded screenshot here :(

File: src/main/java/com/codeborne/selenide/appium/AppiumNavigator.java (hunk 6 above)
Comment (BorisOsipov) [reply src/main/java/com/codeborne/selenide/appium/AppiumNavigator.java:25]: Agree will change it.

**Pull request merged**